README.md

```diff
@@ -17,29 +17,97 @@
 
 ## Installation
 
-Cocotb can be installed by running `pip install cocotb`.
+Cocotb requires:
 
-## Quickstart
+- Python 3.5+
+- A C++ compiler
+- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))
 
-    # Install pre-requisites (waveform viewer optional)
-    sudo yum install -y iverilog python-devel gtkwave
+After installing these dependencies, the latest stable version of cocotb can be installed with pip.
 
-    # Checkout git repositories
-    git clone https://github.com/cocotb/cocotb.git
+```command
+pip install cocotb
+```
 
-    # Install cocotb
-    pip install ./cocotb
+For more details, including how to install a development version of cocotb, see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).
 
-    # Run the tests...
-    cd cocotb/examples/endian_swapper/tests
-    make
+## Usage
 
-    # View the waveform
-    gtkwave waveform.vcd
+As first trivial introduction to cocotb, the following example "tests" a flip-flop.
 
+First, we need a hardware design which we can test. For this example, create a file `dff.sv` with SystemVerilog code for a simple [D flip-flop](https://en.wikipedia.org/wiki/Flip-flop_(electronics)#D_flip-flop). You could also use any other language a [cocotb-supported simulator](https://cocotb.readthedocs.io/en/latest/introduction.html) understands, e.g. VHDL.
 
-## Tutorials and examples
+```verilog
+// dff.sv
+
+`timescale 1us/1ns
+
+module dff (
+    output logic q,
+    input logic clk, d
+);
+
+always @(posedge clk) begin
+    q <= d;
+end
+
+endmodule
+```
+
+An example of a simple randomized cocotb testbench:
+
+```python
+# test_dff.py
+
+import random
+import cocotb
+from cocotb.clock import Clock
+from cocotb.triggers import FallingEdge
+
+@cocotb.test()
+async def test_dff_simple(dut):
+    """ Test that d propagates to q """
+
+    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
+    cocotb.fork(clock.start())  # Start the clock
+
+    for i in range(10):
+        val = random.randint(0, 1)
```

Comment:
It's good practice. You would be surprised how many bugs I've seen because people have assumed "when" they are.
```suggestion
        await RisingEdge(dut.clk)  # Synchronize ourselves with the clock
        val = random.randint(0, 1)
```

Reply:
In this form, the test fails. `ValueError: Unable to resolve to binary >z<` . And I am not sure in any case.

Reply:
```
  6000.00ns ERROR    cocotb.regression                         regression.py:396  in _score_test                     Test Failed: test_dff_simple (result was ValueError)
                                                                                                                     Traceback (most recent call last):
                                                                                                                       File "/examples/dff_simple/test_dff.py", line 20, in test_dff_simple
                                                                                                                         assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
                                                                                                                       File "/cocotb/handle.py", line 409, in __eq__
                                                                                                                         return self.value == other
                                                                                                                       File "/cocotb/binary.py", line 439, in __eq__
                                                                                                                         return self.value == other
                                                                                                                       File "/cocotb/binary.py", line 296, in value
                                                                                                                         return self.integer
                                                                                                                       File "/cocotb/binary.py", line 287, in integer
                                                                                                                         return self._convert_from[self.binaryRepresentation](self._str)
                                                                                                                       File "/cocotb/binary.py", line 193, in _convert_from_unsigned
                                                                                                                         return int(resolve(x), 2)
                                                                                                                       File "/cocotb/binary.py", line 45, in resolve
                                                                                                                         raise ValueError("Unable to resolve to binary >%s<" % string)
                                                                                                                     ValueError: Unable to resolve to binary >z<
  6000.00ns ERROR    cocotb.regression                         regression.py:227  in tear_down                       Failed 1 out of 1 tests (0 skipped)
```

Reply:
The clock hasn't gotten a value yet. This is really more of an issue of `Clock`, which is essentially a driver (just *not*), not running `setimmediate` on the clock signal.

This also highlights my absolute biggest peeve of all about cocotb. The stacktrace there is completely useless. And going in the other direction I sometimes see `truncating bits to match required width` warning messages, but it doesn't tell me on what handle and with what value I tried to do that, or where in my program I am. Sorry to rant in your PR.

Reply:
> The stacktrace there is completely useless

I don't think it's all that bad. It's not too hard (for me) to infer that `dut.q.__eq__` is failing, and that it's because `q` contains a `z`. Perhaps come comments at the end of the lines to aid the user would help, but I don't think the stacktrace is actually missing much information.

Reply:
Hmmm, I just glanced at it, that's different than what we get, but we are pretty far behind. I guess this was recently fixed.

Reply:
"doesn't tell me on what handle and with what value I tried to do that" - at least printing the value would be trivial to add.

Reply:
Reported issue: https://github.com/cocotb/cocotb/issues/1611